students/Thai_H/lesson6/mail_room_part4.py

```diff
@@ -0,0 +1,222 @@
+                                                                                                                 amt = total_donation))
+        thks_letter.close()
+
+#----------------------------------------------------------------------------
+def thank_you_menu_selection():
+    user_selection = input('\n\nTHANK-YOU MENU\nChoose from one of options below:\n\
+        L) List current donors\n\
+        T) Send Thank-you letter (or Add donor and Send one)\n\
+        F) Create/Write Thank-you files\n\
+        M) Back to MAIN MENU\n\
+        Q) Quit the program\n\
+        Please type L, T, F, M, or Q: ')
+    return user_selection
+
+
+#-------------------------------------------------------------------------
+def thank_you_letter():
+    while True:
+        # use of dictionary where 'key' is the menu choice and 'value' is function to call
+        thank_you_menu_function = {
+        'L': create_report,
+        'T': send_thank_you_add_new_donation,
+        'F': create_thank_you_files,
+        'M': main,
+        #'Z': test_create_report,
+        'Q': quit_program
+    }
+        thank_you_selection = thank_you_menu_selection().upper()
+        thank_you_menu_function.get(thank_you_selection, invalid_thank_you_menu)()
+        #invalid_menu is to handle non-of-above choices
+                                                                #
+
+#----------------------------------------
+def sum_donation(d_amt):
+    return sum(d_amt)
+
+
+#-------------------------------------------
+def num_of_donation(n_donation):
+    return len(n_donation)
+
+
+#------------------------------------------
+def avg_donation(total, nd):
+    if nd == 0:
+        raise ValueError('can not divide by zero')
+    return total / nd
+
+
+#--------------------------------------------------------------------------
+def sort_donor_list(unsorted_list):
+    sorted_donation_list = []
+    for d in unsorted_list:
+        total_donation =  sum_donation(unsorted_list[d])
+        number_of_donation = num_of_donation(unsorted_list[d])
+        average = avg_donation(total_donation,  number_of_donation)
+        sorted_donation_list.append([d, total_donation, number_of_donation, average])
+
+    sorted_donation_list = sorted(sorted_donation_list, key = itemgetter(1), reverse = True)
+    return sorted_donation_list
+
+#--------------------------------------------------------------------------
+def sort_by_total_donation(sorted_donation_list):
+    return sorted_donation_list[1]
+
+
+#------------------------------------------------------------------------------
+def create_report():
+    sorted_list = sort_donor_list(list_of_donors)
+    report_header = ['Donor Name', 'Total Donation Amt', 'Num of Donation',  'Avg Per Donation']
+    print('\n\n{:<25}{:>20}{:>20}{:>20}'.format(*report_header))
+    print('-'*90)
+    for donor in sorted_list:
+        print('{:<26}${:>15,.2f}{:>15}{:>25,.2f}'.format(donor[0], donor[1], donor[2], donor[1] / donor[2]))
+    print('\n\n')
```

Comment:
have this return a string (with report created) or a list (with report data) and unit test this function.